modin/core/io/file_dispatcher.py

```diff
@@ -276,14 +274,22 @@ def file_exists(cls, file_path, storage_options=None):
                 else:
                     new_storage_options = {}
 
-                s3fs = S3FS.S3FileSystem(anon=False, **new_storage_options)
+                fs, _ = fsspec.core.url_to_fs(file_path, **new_storage_options)
                 exists = False
                 try:
-                    exists = s3fs.exists(file_path) or exists
-                except (NoCredentialsError, PermissionError, EndpointConnectionError):
-                    pass
-                s3fs = S3FS.S3FileSystem(anon=True, **new_storage_options)
-                return exists or s3fs.exists(file_path)
+                    exists = fs.exists(file_path)
+                except (
+                    NoCredentialsError,
+                    PermissionError,
+                    EndpointConnectionError,
+                    ConnectTimeoutError,
+                ):
+                    fs, _ = fsspec.core.url_to_fs(
+                        file_path, anon=True, **new_storage_options
+                    )
+                    exists = fs.exists(file_path)
+
+                return exists
         return os.path.exists(file_path)
```

Comment:
I'm confused about when we hit this line. I think it's when `isinstance(file_path, str)` is False? Will `os.path.exists` ever return True in that case? If we do want this `os.path.exists` line still, can we move it up and return early if it's True? that will make the code less indented. https://blog.codinghorror.com/flattening-arrow-code/

Reply:
I've changed if-else branching to see when each flow happens. Please, take a look again

Reply:
Can we do `if not isinstance(file_path, str): return os.path.exists(file_path)` and then dedent the code that's currently under the if clause? I think the code will be easier to read if it's less indented.

Reply:
> Can we do `if not isinstance(file_path, str): return os.path.exists(file_path)` and then dedent the code that's currently under the if clause? I think the code will be easier to read if it's less indented.

But we do `return os.path.exists(file_path)` not only in case `if not isinstance(file_path, str)`. With your suggestion this block will look like:
```python
if not isinstance(file_path, str):
    return os.path.exists(file_path)`
elif isinstance(file_path, str) and fsspec.core.split_protocol(file_path)[0] in _SUPPORTED_PROTOCOLS:
   # protocols processing stuff
else:
    return os.path.exists(file_path)` # processing the last case when `file_path` just a local filesystem path
```

Reply:
Can we do this:

```python
if not isinstance(file_path, str) or fsspec.core.split_protocol(file_path)[0] not in _SUPPORTED_PROTOCOLS:
  return os.path.exists(file_path)
# protocols processing stuff not inside "if" or "else"
```

Reply:
I understand what you mean! We can do this. I've applied this for `FileDispatcher.file_exists` and `CsvGlobDispatcher.file_exists/get_path`